thorn/src/kernel/mailbox.c

```diff
@@ -1,52 +1,56 @@
 #include "kernel/mailbox.h"
 
-bool mailbox_request (volatile unsigned int * data_ptr, channel_t channel) {
-    volatile unsigned int * buffer = data_ptr;
-
+bool mailbox_request (volatile unsigned int * buffer, channel_t channel) {
     if (DUMP_BUFFER) {
         LOG ("Dumping message buffer before sending:");
         int_dump ((unsigned int *) buffer);
         hex_dump ((byte_t *) buffer);
     }
 
-    // 28-bit address (MSB) and 4-bit value (LSB)
-    unsigned int outgoing = ((unsigned int) ((long) data_ptr) & ~0xF) | (channel & 0xF);
-
-    // Wait until we can write
-    while (get32 (MBOX0 + MBOX_STATUS) & MBOX_WRITE_FULL)
-        ;
+    mailbox_message_t message = {buffer, channel};
 
-    // Write the address of our buffer to the mailbox with the channel appended
-    put32 (MBOX0 + MBOX_WRITE, outgoing);
-
-    unsigned int incoming;
-    while (1) {
-        // Is there a reply?
-        while (get32 (MBOX0 + MBOX_STATUS) & MBOX_READ_EMPTY)
-            ;
-
-        incoming = get32 (MBOX0 + MBOX_READ);
-
-        // Is it a reply to our message?
-        if (outgoing == incoming) {
-            if (buffer[1] != MBOX_SUCCESS) {
-                if (DUMP_BUFFER) {
-                    LOG ("Dumping failed message buffer:");
-                    int_dump ((unsigned int *) buffer);
-                    hex_dump ((byte_t *) buffer);
-                }
-
-                return false;
-            }
-            break;
-        }
-    }
+    mailbox_write (&message);
+    mailbox_read (&message);
 
     if (DUMP_BUFFER) {
         LOG ("Dumping successful message buffer:");
         int_dump ((unsigned int *) buffer);
         hex_dump ((byte_t *) buffer);
     }
 
-    return true;
+    if (buffer[1] == PROPERTY_RESPONSE_SUCCESS) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+unsigned int mailbox_message_to_register_value (mailbox_message_t * message) {
+    return ((unsigned int) ((long) message->data) & ~0xF) | (message->channel & 0xF);
```

Comment:
As far as I can tell, what you try to do here is to merge 28 bits of data with 4 bits of channel into the 32 bit message the mailbox expects. I.e., you try to combine `message` into this format:
```
|            data            |chnl|
|^^^^^^^^^^^^^^^^^^^^^^^^^^^^|^^^^|
|-------------28-------------|--4-|
```

Now, have a look at the structure of `mailbox_message_t`:
```c
typedef struct {
    unsigned int data    : 28;// most significant bits contain shared memory address
    channel_t    channel : 4; // least four significant bits contain channel
} mailbox_message_t;
```
Incidentally, this structure is 4 bytes long and gets packed into this format:
```
|            data            |chnl|
|^^^^^^^^^^^^^^^^^^^^^^^^^^^^|^^^^|
|-------------28-------------|--4-|
```
That means the data is already in the proper format; all we have to do now is tell the compiler that it is what it should be:
* Casting (assuming you pass `message` by value, otherwise omit the `&`):
```suggestion
    return * (unsigned int *) & message;
```
* Bit-wise OR:
```suggestion
    return (unsigned int) (message->data << 4) | message->channel; // no need to specifically clear bits
```

Reply:
I just realised this issue was caused by me not communicating clearly how `mailbox_message_t` should be used - 
`message.data` is supposed to store the **28 most significant bits** of the memory address after it has been truncated to a 32bit integer. We know the **4 least significant bits** are 0 anyways. I should've made that clear and provide an appropriate Initialise function.

Reply:
I changed the source to pass `mailbox_message_t` by value. Unfortunately, the cast you suggested compiled, but didn't work on the running machine. I had errors on the console. Therefore, I simplified it to `return (message.data & ~0xF) | message.channel;`.

Reply:
It is dependent on this change: https://github.com/anarchuser/rose/pull/63/files/d06e110f36f681d0931ad45cffad47b15c71abb9#r767591035

Reply:
I saw that, but see my comment below. As soon as I do the cast like this the boot hangs at `src/common/rng.c:15:[0] Initialised RNG successfully`.

Reply:
hm oaky, I'll have a deeper look

Reply:
The current version only casts but doesn't shift. Is that intended?
https://github.com/anarchuser/rose/blob/c2dea750a809d538f11a8cb7fe1615508e756a26/thorn/src/kernel/mailbox.c#L10

Reply:
as for why it hangs is because if data doesn't point to a proper message, the mailbox doesn't do anything

Reply:
Yes, because otherwise the above mentioned crash happens. In the function where I cast the mailbox message for the register I zero out the 4 Bit.

Reply:
so you get the crash if 1) you shift and 2) you do not zero out the 4 LSB in the register function?

Reply:
Correct.

Reply:
okay, then leave it be for now. Maybe I'll have a look at it tomorrow, the error is probably more subtle